### PR TITLE
[Home] Search button and functionality

### PR DESCRIPTION
**Pull request by OliRivDev**
* Search Button linked to a new ViewController
* Search functionality implemented
* Cancel to dismiss ViewController and go to Home
* Click name in result to go to details ViewControlle

### CODE REVIEW

File: MarvelCharsApp/MarvelCharsApp/ViewModels/SearchViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        } else {
+            searchedItems = allCharacters
+        }
+    }
+    
+}
+
+extension SearchViewModel: MarvelRepositoryDelegate {
+    func didFailFetching(error: Error) {
+        print(error)
```
Comment (jmacboy) [inline MarvelCharsApp/MarvelCharsApp/ViewModels/SearchViewModel.swift]: this should reflect on the UI with an alert or some kind of message. (change a variable here, reflect there)

File: MarvelCharsApp/MarvelCharsApp/Views/SearchTableViewController.swift [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    }
+    
+    // MARK: - Table view data sourc
+    
+    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return searchVm.count()
+    }
+    
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: "nameCell", for: indexPath)
```
Comment (jmacboy) [inline MarvelCharsApp/MarvelCharsApp/Views/SearchTableViewController.swift]: Use identifiers as constants, not just strings

File: MarvelCharsApp/MarvelCharsApp/Views/SearchTableViewController.swift [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        let character = searchVm.get(byIndex: indexPath.row)
+        cell.textLabel?.text = character.name
+        return cell
+    }
+    
+    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let vc = UIStoryboard(name: "Main", bundle: nil)
```
Comment (jmacboy) [inline MarvelCharsApp/MarvelCharsApp/Views/SearchTableViewController.swift]: same here for both identifiers, also if Main storyboard is used somewhere else, create a constant in an external class

File: MarvelCharsApp/MarvelCharsApp/Views/SearchTableViewController.swift [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            .instantiateViewController(withIdentifier: "DetailsViewControllerID") as! DetailsViewController
```
Comment (jmacboy) [inline MarvelCharsApp/MarvelCharsApp/Views/SearchTableViewController.swift]: never force unwrap, is best to use guard let or validate for nils.

File: MarvelCharsApp/MarvelCharsAppTests/SearchViewModelTests.swift [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    
+    
+    func prepareAndWaitFirstFetchExpectation() {
+        let expectationForFecthAll = self.expectation(description: "Request to API")
+        search.refreshData = {
+            self.refreshCallsCount += 1
+            expectationForFecthAll.fulfill()
+        }
+        search.getAllcharacters()
+        self.waitForExpectations(timeout: 2, handler: nil)
```
Comment (jmacboy) [inline MarvelCharsApp/MarvelCharsAppTests/SearchViewModelTests.swift]: unit tests must never use a real API, use the mocks here and remove the wait.

File: MarvelCharsApp/MarvelCharsAppTests/SearchViewModelTests.swift [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    
+    func testGetCharacterByIndex() throws {
+        let items = search.get(byIndex: 0)
+        XCTAssertNotNil(items)
+    }
+    
+    func testGetCharacterByCorrectName() throws {
+        prepareSearchlocalExpectation()
+        search.getCharacters(byName: "Spider Man")
+        self.waitForExpectations(timeout: 2, handler: nil)
```
Comment (jmacboy) [inline MarvelCharsApp/MarvelCharsAppTests/SearchViewModelTests.swift]: is the same in all your tests.

**Review by jmacboy — CHANGES_REQUESTED** [review]

File: MarvelCharsApp/MarvelCharsApp/ViewModels/SearchViewModel.swift (hunk 1 above)
Comment (OliRivDev) [reply MarvelCharsApp/MarvelCharsApp/ViewModels/SearchViewModel.swift]: Change to a errorHandler closure

File: MarvelCharsApp/MarvelCharsApp/Views/SearchTableViewController.swift (hunk 2 above)
Comment (OliRivDev) [reply MarvelCharsApp/MarvelCharsApp/Views/SearchTableViewController.swift]: I did a static class to access to identifiers directly

File: MarvelCharsApp/MarvelCharsApp/Views/SearchTableViewController.swift (hunk 3 above)
Comment (OliRivDev) [reply MarvelCharsApp/MarvelCharsApp/Views/SearchTableViewController.swift]: Resolved

File: MarvelCharsApp/MarvelCharsAppTests/SearchViewModelTests.swift (hunk 6 above)
Comment (OliRivDev) [reply MarvelCharsApp/MarvelCharsAppTests/SearchViewModelTests.swift]: Changed to mockAPI

File: MarvelCharsApp/MarvelCharsAppTests/SearchViewModelTests.swift (hunk 5 above)
Comment (OliRivDev) [reply MarvelCharsApp/MarvelCharsAppTests/SearchViewModelTests.swift]: Changed to mockAPI

File: MarvelCharsApp/MarvelCharsAppTests/SearchViewModelTests.swift [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    
+    func testGetCharacterByEmptyName() throws {
+        search.getCharacters(byName: "")
+        let count = search.count()
+        XCTAssertEqual(2, refreshCallsCount)
+        XCTAssertEqual(3, count)
+    }
+}
+
+//Not using mock in this test
```
Comment (jmacboy) [inline MarvelCharsApp/MarvelCharsAppTests/SearchViewModelTests.swift]: I think you're using it this time, remove this comment and you're good to go

**Review by jmacboy — CHANGES_REQUESTED** [review]

**Review by jmacboy — APPROVED** [review]

**Pull request merged**